File: 410.py
class Solution:
    def splitArray(self, nums, m):
        N = len(nums)
        if m >= len(nums):
            return max(nums)
        dp = [[float('inf')] * N for _ in range(m + 1)]
        from itertools import accumulate
        s = [*accumulate(nums, lambda x, y: x + y)]
        dp[1] = s
        sum = [[s[j] - s[i] for j in range(N)] for i in range(N)]
        for i in range(2, m + 1):
            for j in range(i - 1, N):
                dp[i][j] = min(max(dp[i - 1][k], sum[k][j]) for k in range(i - 2, j))
        return dp[-1][-1]


if __name__ == '__main__':
    solution = Solution()
    print(solution.splitArray([7, 2, 5, 10, 8], 2))

File: 410.kt
class Solution {
    fun splitArray(nums: IntArray, m: Int): Int {
        val N = nums.size
        if (m >= N) {
            return nums.max() ?: 0
        }
        val dp = Array(m) { IntArray(N, { Int.MAX_VALUE }) }
        var index = 0
        nums.fold(0, { acc: Int, n: Int ->
            val result = acc + n
            dp[0][index] = result
            index += 1
            result
        })
        for (i in 1 until m) {
            for (j in i until N) {
                for (k in i - 1 until j) {
                    dp[i][j] = minOf(dp[i][j], maxOf(dp[i - 1][k], dp[0][j] - dp[0][k]))
                }
            }
        }
        return dp.last().last()
    }
}

fun main(args: Array<String>) {
    val solution = Solution()
    println(solution.splitArray(intArrayOf(7, 2, 5, 10, 8), 2))
}
